### ISSUE TITLE
enable migrating an asset to a new key

### ISSUE THREAD

**Issue by sryza** [dagster-io/dagster#9512]
If you rename table "x" to "y", you probably want to change the asset key that represents it in Dagster.

We could expose a CLI for this. E.g.

```
asset move --from a>b>c --to x>y>z
```

Requests:
- https://dagster.slack.com/archives/C01U954MEER/p1660918331045069

**Comment by sryza** [comment]
Thoughts on this @prha ?

**Comment by simonvanderveldt** [comment]
Seems to be the same/a duplicate of https://github.com/dagster-io/dagster/issues/3472

My current understanding is that the ID of an asset is it's name, this is not great, because that means changing the name means a new asset and losing the existing history.
Wouldn't it be better to (be able to) assign a unique ID (let's say a UUID) to an asset separately from the name and use that ID for all references in the code/database. Then renames would have no impact at all.

P.S. Airflow suffers from the same issue, it makes renames a pain, so renames don't happen/get postponed so in practice you're left with stuff with the wrong/incorrect name because renaming is just too much effort.

**Comment by stkbailey** [comment]
Going to add to this that my desired funcitonality is something akin to `terraform state mv` to migrate partition sets from one asset to another. See this [Slack message](https://dagster.slack.com/archives/C01U5LFUZJS/p1673273771633349)

One thought your comment @simonvanderveldt , I don't really care what happens under the hood, but I do want the ability to manage and maintain the "addresses" of all my assets for the global graph. We are building a multi-location asset graph: dbt models are maintained next the dbt project, most dagster jobs are maintained in another repo, but they have cross-repo dependencies. 

So the global address (and not an internal id) is the way I can coordinate across these different repositories -- i.e. `AssetKey(["snowflake", "db", "schema", "table"])` is meaningful across all these repositories, and I wouldn't want a reference to that to fail because it was pointing to different internal IDs. Not really sure if that's what you're talking about.

**Comment by prha** [comment]
Yeah, I think the tricky thing about this is figuring out where to resolve the public address (asset key), to some sort of internal storage representation so that we can easily do these types of operations without having to mutate all event history for all time.  It's especially tricky since we want to limit the number of lookups we need to do in order to write events into the storage layer.

**Comment by deoqc** [comment]
I imagine 2 solutions:
1) Hard-code and id in the `@asset` decorator, just as we do w/ asset key, and be able to use this as the identity of the asset.
2) Generate automatically a config w/ these asset id's (some text-format w/ asset_key --> asset_id mapping) so that we can commit to version control and can easily migrate.

Solution 1) is less user-friendly at first but allows for ease migration (just update the key). Solution 2) is more user friendly at first but requires more sync when doing renaming (needs to sync updating the asset key and the mapping).

As a final option, you could have both at same time, w/ 2) the default and 1) an option that takes precedence.

**Comment by the4thamigo-uk** [comment]
Was this ever implemented?

**Comment by sryza** [comment]
@the4thamigo-uk this has not yet been implemented

**Comment by AlanPedro** [comment]
This would be super useful

**Comment by brandonpeebles** [comment]
+1 to this. Would be really helpful for data migrations or just rethinking our naming conventions in Dagster as we build things out. 

Maybe this could involve some human input where we tell Dagster some "rule" or "pattern" by which we changed the asset keys? E.g. we're moving from `my/asset/key` to `my/new_asset/key` we tell Dagster that we're appending "new_" to assets that match some pattern. Of course, an automatic solution would be better — but really anything that enables the migration ability at all would be helpful.

**Comment by DennisSchwartz** [comment]
Giving another 1+ to this.
I have an asset which takes a lot of resources to materialise and I want to rename it without re-processing all the data.

I'm considering a workaround where I would report materialisation events via the API to copy over the events of the old asset. But this will cost a lot of Dagster tokens. (My asset has many partitions)

**Comment by geo909** [comment]
We're going through a big refactor, and this is really needed, we'll have to start clean for many of our assets now.

I was wondering where this stands right now. Is there any hope it will be implemented in the near future?

**Comment by TCronino** [comment]
+1 to this - multiple instances of accidentally rerunning hundreds of partitions after assets getting renamed
